xpad.c

```diff
@@ -952,6 +961,36 @@ static void xpadone_process_packet(struct usb_xpad *xpad, u16 cmd, unsigned char
 				 (__u16) le16_to_cpup((__le16 *)(data + 8)));
 	}
 
+	/* paddle handling */
+	/* based on SDL's SDL_hidapi_xboxone.c */
+	if (xpad->mapping & MAP_PADDLES) {
+		/* HACK: Paddle data is delivered differently for different models/firmwares. There's probably a better way to detect this... */
+		if (xpad->name == "Microsoft X-Box One Elite pad"){
```

Comment:
I was actually fairly nervous about hardcoding a "magic value" like this, particularly because there may someday be copycat controllers that use the same packet format. I also notice that the rest of this driver has avoided directly checking `xpad->name`. Is that by intention? 

I _originally_ wanted to do what SDL does and check the size of the packet as a detection heuristic, but it seems like this driver just allocates the full 64 bytes regardless of actual packet size.

This approach will be a particular problem when we eventually try implementing handling of additional packet formats, since that can't be detected with this method. Without a size heuristic, I imagine this would require GIP?

Reply:
yes, you should rather check idVendor & idProduct

Reply:
also check `dmesg` after you attach your gamepad. The firmware version might be in `bcdDevice`

Reply:
@paroj Yeah, I'd suggest that, too. At least for HID mode, we can reliably decide about the packet format by looking at the PID. There are some copycat controllers which don't use the rumble programming in the same way, and we can add a quirk flag for that by looking at the MAC OUI - apparently, that's not possible for USB.

@chaorace Did you check if the PID changes with the new firmware? If yes, that should be enough. Otherwise try to get the firmware version from the USB descriptor (I don't know how to do that but @paroj suggests `bcdDevice`).

I cannot say much about the 5.13 firmware for XB1S controllers yet in HID mode (I didn't receive it yet). With later firmwares, HID and GIP are more identical, and GIP is mostly the USB protocol, I think. So once I got that firmware, you could look at xpadneo commits (in my personal fork where I prepare stuff, or in the main atar-axis repo). From what I found with the older firmware, the paddles are only clearly visible as separate bits in profile 1, 2, or 3. Profile 0 just matches them with A,B,X,Y.

I don't think that looking at the packet size is a good idea, that's actually what makes SDL nullifying our HID fixups in xpadneo (I'm patching the raw HID packets to match a common format but I cannot change the size at that point). SDL can match Bus/VID/PID/Version but it seems to totally ignore that if it detects an authentic Xbox controller in native HID mode.

And I must admit that I never tried this xpad driver but I'm sure it can be useful to share some thoughts and encountered problems, which is what I wanted to do with my comments.

Reply:
Looks like `bcdDevice` does the trick! On v4.8.1908.0, I get a value of `4.08` for the bcdDevice value. On v5.13.3143.0, I get a value of `5.0d`. The major version number is pretty obvious, though I'm as of yet not quite sure how to interpret the minor version scheme that MS is using within the descriptor.

By the way, the PID remains the same on both firmwares for my XBE2: `0x0b00`

> From what I found with the older firmware, the paddles are only clearly visible as separate bits in profile 1, 2, or 3. Profile 0 just matches them with A,B,X,Y.

That's interesting, do you recall which hardware/software configuration that was? In slot 0 on v4, my XBE2 reports paddles entirely independently without lighting up any other inputs.

> I don't think that looking at the packet size is a good idea, that's actually what makes SDL nullifying our HID fixups in xpadneo

That makes sense. It's still a tempting option, though, since it takes away the painful the burden of meticulously documenting the history of minor closed-sourced firmware releases :smirk:. Oh well... better to do it the right way rather than rely on a heuristic.

Anyways... after upgrading back to v5.13.3143.0, I am unfortunately still not able to figure out the paddles. I dug a little deeper into the SDL source files and found [this concerning excerpt](https://github.com/libsdl-org/SDL/blob/9125b244e7679a4641af02626baf01dc53e2dc6d/src/joystick/hidapi/SDL_hidapi_xboxone.c#L679-L681), which seems to imply that the paddle data has indeed been dropped in later v5 firmwares.

Reply:
> I get a value of 5.0d. The major version number is pretty obvious, though I'm as of yet not quite sure how to interpret the minor version scheme that MS is using within the descriptor.

```
>>> 0xd
13
```

Reply:
Doh. :laughing:

Reply:
I don't think the minor version is important for anything handling protocol differences. The differences are in x.y, not x.y.z.whatever. The major differences are between v4 and v5, at least in HID mode this is obvious: v4 is legacy Bluetooth, v5 is BLE. There's a special pairing mode of older firmwares that shows v9 but that seems to be gone, it was activated when the controller detected an Android-like OS as the host (thus, it did this on Linux). v9 is not relevant for USB connections.

Then there are differences between 5.x version but those should be quite minor. In HID mode, we may see different PIDs per different "x", and maybe small protocol differences.

Regarding the pedals, this was with an early firmware version of the XBE2 controller back when it had the "each button is reported twice and the report is huge" bug. I didn't look into it very detailed later but I found that pressing the pedals in profile slot 0 always activated the A,B,X,Y bits at the same time. I'm somehow under the impression that there is a programming interface to set in-controller memory for what key codes the paddles should send (some of the undocumented vendor extensions in the HID descriptor, one may actually be an RGB interface for the power LED face button). Here's some documentation about the different models and the reports, the report layout for USB/GIP should be very similar: https://github.com/atar-axis/xpadneo/tree/master/docs/descriptors

About the concerning excerpt: I found the same differences in xpadneo but didn't need to use the packet length as a discriminator - xpadneo purely uses the HID descriptor and PID to detect the protocol layout. The XBE2 controller was the first model to introduce non-gaming buttons to a new sub-device, seems they have given up the idea with v5 firmwares. It was probably meant to work around a bitmap problem with button positions in Linux (which has BTN_MODE in a position incompatible with early Xbox controllers). In HID mode, the sub-device is quite obvious from the descriptor: both reports belong to different device applications. In GIP/USB mode, it's just two different reports. You can simply map those to the same button, you won't ever see multiple variants at the same time, it's either one variant or another.

The solution SDL took seems somewhat "concerning" because they don't ever look at the HID descriptor, and thus the packet size is mostly the only thing they can work with (there's another one but that doesn't integrate well with the design of the code). You should avoid walking the same path.

There's only one instance where we actually look at the report size not just for boundary checks:
* https://github.com/atar-axis/xpadneo/blob/4fd620cd6cb80fb0c1490dc1c864108679d91ab1/hid-xpadneo/src/hid-xpadneo.c#L862

That instance actually handles the lock position for each trigger and the profile slot switcher. It shifted position between updates without changing the PID of the controller.

Some early versions of the XBE2 firmware switched packet sizes by 1 byte without any real difference in the bitmaps. AFAIR, the only difference was moving the Guide button to a different sub-device. To me this means: Better not try to second-guess the packet format by packet size.

Reply:
Oh, your "doh" just opened my eyes: In version "x.y" you mean "y" as the minor version - not the components following it. But my comment is still mostly true, with different y in "5.y", there's not much difference usually - at least in HID mode. The difference in HID mode is that the version would be `0x05 0x13` while in USB it seems `0x05 0x0d`.

Reply:
I tried spying on how Windows works with v5 using Wireshark earlier today. I noticed that the GIP packets do indeed look very similar to the HID packets that we get in Linux, including the lack of any paddle data in the main packet.

I dug a little deeper and found that the missing paddle input data was accessible via a different packet type whose header starts with `0x0c`. Unfortunately, I couldn't manage to get my controller to send these `0x0c` packets when hooked up to my Linux machine.

I'm not sure what to make of that... maybe there's a set of special init bits that will make the controller send the extra packets? Or perhaps this has something to do with "Linux mode"?

Reply:
The "init" part you suggest may be interesting... Maybe the controller will be initialized with some special data to program the firmware how to send the paddle bits. The xone project may have some progress here as we discovered that there is a TLS handshake during setup of the wireless GIP session which is needed to obtain some extended functionality. But as far as I can remember, this only affected accessory devices like headsets. I wrote a simple GIP decoder for wireshark, if you'd like to use it, join the xpadneo Discord server and PM me.

Reply:
I'd also recommend checking out https://github.com/medusalix/xone but I might be slightly biased in that regard.
`xpad` doesn't properly initialize GIP devices, requiring a ton of quirks for different controller models.
`xone` performs the full GIP handshake (minus the encryption part) and adding support for the `0x0c` packets should be no big deal.

Reply:
@medusalix I've actually been referencing your repository quite a bit! Your work RE work with the GIP packet and [command bits](https://github.com/medusalix/xone/blob/52eb2ff02fb6b21bd0815f4ea7460a76ad2b3c67/bus/protocol.c#L26-L45) in particular has been a huge boon to me -- thank you **very much** for that. To clarify a bit further: the GIP packet type that I spied carrying the paddle data is currently documented as `GIP_CMD_FIRMWARE` (_not sure if that makes "FIRMWARE" a misnomer, considering the packet contents observed in my testing?_)

Regarding initialization: this has been a major worry of mine. I don't think it makes sense to try and restructure xpad to speak GIP at this point (especially since xone exists), so my hope is to hack it out with the driver's existing init bit system.

The first showstopper issue regarding this is that I have no idea how the existing init bits were originally RE'd! I happened to notice (after the initial wave of identification packets) that the Windows driver is sending a pair of `0x05` type GIP packets which look suspiciously similar to the init bits that xpad was already using for One S controllers. Unfortunately, it didn't seem to have any effect when I swapped these into xpad's init bits... do you happen to have any knowledge about what these packets are encoding?

Packet 1:
`0x05, 0x20, 0x02, 0x0f, 0x06, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x55, 0x53, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00`

Packet 2:
`0x05, 0x20, 0x03, 0x01, 0x00`

Reply:
You should be able to compare the different drivers with the Wireshark GIP dissector I've sent to you. The names within the dissector are based on the xone driver.

Reply:
That's not a bad idea... if the initialization process changed between v4 and the current driver, maybe comparing the two will reveal some additional information?

After work today I'll run a capture of both firmwares initializing. I'll also share gist links for both wireshark capture dumps -- just in case you folks are interested in taking a peek.

Reply:
@chaorace There's a bit in the `options` field of the GIP header which denotes whether a command is part of the core protocol or a vendor-specific command. The `0x0c` command you're getting most likely doesn't have that bit set, which means it's device-specific and not `GIP_CMD_FIRMWARE`.

Reply:
@medusalix Yes, that difference is also visible in the dissector, @chaorace I suggest you take a look at the LUA source code of the dissector to get a better idea of the protocol. The specifications within are probably incomplete so I'd appreciate any new ideas. The dissector uses "GIP" for protocol specific stuff, and "ACC" for accessories stuff, but that naming is not finalized.

Reply:
Found it :grin: 
`0x4d, 0x10, 0x01, 0x02, 0x07, 0x00`

I tested that one just now as an extra set of init bits and the extra packets started sending.

So... it turns out that I was missing something extremely important: By default, the controller does *not* send the extra `0x0c` packets -- not even on Windows! It's the _Accessories app_ that tells the controller to start sending these extra packets, **not the driver**. You can see from the excerpt below the exact moment that the second packet type is turned on by the Accessories app (_captured by spinning the control stick with one hand while opening the accessories app with the other_).

This means that @kakra was right on the money. There is indeed a packet that can be used to modify the controller behavior at runtime. Several more of these seem to get fired off when opening Accessories... so who knows what else it's doing? I'll share the full dump shortly once I've had a chance to push my changes.

```
No.     Time           Source                Destination           Protocol Length Info
   2077 34.160533      1.1.2                 host                  GIP      77     ACC INPUT <--- Normal 1 packet input data behavior

0000  1b 00 60 4a fa 63 0b 83 ff ff 00 00 00 00 09 00   ..`J.c..........
0010  01 01 00 01 00 82 01 32 00 00 00 20 00 5a 2e 00   .......2... .Z..
0020  00 00 00 00 00 90 8f bd 4e 1f 01 83 fc 00 00 00   ........N.......
0030  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00   ................
0040  00 00 00 00 00 0e 28 3d 02 a2 2b 3d 02            ......(=..+=.

No.     Time           Source                Destination           Protocol Length Info
   2079 34.162952      host                  1.1.2                 USB      33     URB_INTERRUPT out <--- 0x4d packet fired by Accessories

0000  1b 00 00 6b fb 63 0b 83 ff ff 00 00 00 00 09 00   ...k.c..........
0010  00 01 00 01 00 02 01 06 00 00 00 4d 10 01 02 07   ...........M....
0020  00                                                .

No.     Time           Source                Destination           Protocol Length Info
   2081 34.168510      1.1.2                 host                  GIP      40     GIP ACKNOWLEDGE

0000  1b 00 20 6b 93 67 0b 83 ff ff 00 00 00 00 09 00   .. k.g..........
0010  01 01 00 01 00 82 01 0d 00 00 00 01 20 01 09 00   ............ ...
0020  4d 00 02 00 00 00 00 00                           M.......

No.     Time           Source                Destination           Protocol Length Info
   2083 34.172473      1.1.2                 host                  USB      33     URB_INTERRUPT in

0000  1b 00 00 7b 88 64 0b 83 ff ff 00 00 00 00 09 00   ...{.d..........
0010  01 01 00 01 00 82 01 06 00 00 00 4d 00 01 02 07   ...........M....
0020  00                                                .

No.     Time           Source                Destination           Protocol Length Info
   2087 34.176530      1.1.2                 host                  GIP      77     ACC INPUT <--- Same packet from normal input data behavior

0000  1b 00 10 00 8b 5d 0b 83 ff ff 00 00 00 00 09 00   .....]..........
0010  01 01 00 01 00 82 01 32 00 00 00 20 00 5b 2e 00   .......2... .[..
0020  00 00 00 00 00 00 80 88 2d 1f 01 83 fc 00 00 00   ........-.......
0030  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00   ................
0040  00 00 00 00 00 90 66 3d 02 bd 6a 3d 02            ......f=..j=.

No.     Time           Source                Destination           Protocol Length Info
   2089 34.180790      1.1.2                 host                  USB      48     URB_INTERRUPT in <--- Extra packet with additional input data!!!

0000  1b 00 60 4a fa 63 0b 83 ff ff 00 00 00 00 09 00   ..`J.c..........
0010  01 01 00 01 00 82 01 15 00 00 00 0c 00 01 11 00   ................
0020  00 00 00 00 00 00 80 88 2d 1f 01 83 fc 00 00 0a   ........-.......
```

Reply:
As promised, here are the wireshark dumps:

[Elite Controller v4.8.1908](https://drive.google.com/file/d/1y_H4FhJMHnpZiEDjtHjaQQQnyeCch_8Q/view?usp=sharing) & Accessories app v300.2202.25001.0

[Elite Controller v5.13.3143.0](https://drive.google.com/file/d/1A1Qk828NAL2joaQnNVX0VuWaVoiiuhmS/view?usp=sharing) & Accessories app v300.2202.25001.0

Reply:
@chaorace Thanks for your work. Maybe we should add support for the `USB_INTERRUPT out/in` packets as well? Do you know what it is and feel confident to add that to the lua script?

Reply:
@kakra Absolutely, I'd love to... I actually tried to improve the script a little earlier today, but my Lua is just so rusty now that I completely failed. I'm going to try again from scratch tomorrow after getting some sleep.

By the way, it would really help me out if you could set up a git repo for it (even a private one would be fine!) It's a lot easier for me to recover from mistakes and discuss potential issues with the help of version control.

Reply:
@chaorace I've sent an invitation to https://github.com/xpadneo/gip-dissector